### ISSUE TITLE
Add ability to stop Input

### ISSUE THREAD

**Issue by cretz** [dnstap/golang-dnstap#27]
Currently this is not very usable as a library since `FrameStreamSockInput`s `ReadInto` runs forever, leaking the goroutine. Consider a Close on the input interface (I may submit a PR if I end up using this as a library).

Also, in general this code logs errors silently among other practices that are poor for library use. If adapting the code for use as a library is not a goal, please at least add caveats via comments to warn other potential users.

**Comment by cmikk** [comment]
Agreed. The golang-dnstap API needs that and other refactoring to be more library-like. Discussion in #23 also noted the need for a zero-copy (or at least less-copying) approach for higher-volume dnstap processors.

As a start, I'd propose:
  * expose lower-level Reader/Writer APIs in addition to the channel-based Input/Output, for callers which need more control
  * add "logger" callbacks to the FrameStreamFOOInput/Output structures
  * add Decoder/Encoder APIs which speak Dnstap structures instead of byte buffers

Other proposals and/or contributions welcome.

**Comment by cmikk** [comment]
Feedback welcome on the API changes in: https://github.com/dnstap/golang-dnstap/tree/issue-27

Note that this requires the "reader-writer" branch of https://github.com/farsightsec/golang-framestream

**Comment by cretz** [comment]
I ended up using the framestream lib directly. Didn't dig too deep, but might want to golint (first word of Godoc invalid), also seeing misspellings like "SocktWriter", etc. But I'll defer to others on substantive review.

**Comment by mvanholsteijn** [comment]
Just added a new output destination which allows dnstap to sent messages to AWS S3 buckets; This issue prevents dnstap itself gracefully shutdown all of the output streams. 

This issue just celebrated its 3rd anniversary: Is there any chance that somebody is going to pick this up?
